### DISCUSSION TITLE
Allow for a nonce to be added to livewire and filament (inline) scripts to improve CSP security

### DISCUSSION THREAD

**Discussion by Roboroads** [filamentphp/filament#7032] [Ideas]
Currently, while livewire supports it, there is no nonce support in filament admin. This makes that we have to add `script 'unsafe-inline'` to our CSP, which defeats the purpose of a CSP.

**Comment by danharrin** [comment]
I have no experience with this, so please PR as long as it doesn't hinder the maintainability of the framework.

**Reply by Roboroads** [reply]
Sure!

**Comment by ghost** [comment]
This is very much needed. Can't use Filament now due to all the inline stuff being blocked by a proper CSP.

See: https://content-security-policy.com/nonce/

**Comment by romansh** [comment]
Is there any plan to implement CSP compatibility?

**Reply by danharrin** [reply]
I do not have any plans at the moment, Livewire does not even support it yet AFAIK, only raw Alpine. And it would greatly decrease the developer experience for us.

**Comment by rossbearman** [comment]
@danharrin While Livewire doesn't support `unsafe-eval`, and is unlikely to in the future, it does support safe inlines by passing a nonce to `@livewireScripts` and `@livewireStyles`, since a PR [back in 2020](https://github.com/livewire/livewire/pull/1156). This negates the requirement for `unsafe-inline` in the CSP.

`@livewireScripts(['nonce' => $nonce])`
`@livewireStyles(['nonce' => $nonce])`

It's now managed through [FrontendAssets](https://github.com/livewire/livewire/blob/fd7ddd1dd043f700d2f5ec0d708c163415555915/src/Mechanisms/FrontendAssets/FrontendAssets.php#L177) in Livewire.

It should be possible to remove the need for `unsafe-inline` in Filament if we were able to pass a nonce through to Livewire on every request and have it appended to any inline scripts or styles Filament uses on top (for example in [assets.blade.php](https://github.com/filamentphp/filament/blob/3.x/packages/support/resources/views/assets.blade.php)). 

The biggest potential issue I see with it would be any plugins that currently inline scripts or styles would need to add support for it, or they'd be blocked if `unsafe-inline` was removed.

**Reply by danharrin** [reply]
Thanks for the info! Would accept a PR :)

**Reply by rossbearman** [reply]
Great.

My thinking would be to add a new panel method `cspNonce()` that accepts a closure or a string, to allow the developer to pass in some kind of generator.

We'd then have an additional helper, perhaps `Filament\Support\get_csp_nonce()` that returns an empty string if one isn't configured, or evaluates the closure and returns `nonce="foo"`.

The biggest change would then be replacing every `<script>` or `<style>` tag in the Filament source with `<script {{ \Filament\Support\get_csp_nonce() }}>` and encouraging plugin developers to do the same with anything they're inlining.

For an end user using Spatie's CSP package, enabling safe inline would then just be a case of adding `->cspNonce(fn (): string => csp_nonce())` to the panel definition.

Does that seem like an appropriate approach for Filament? `unsafe-inline` tends to be scored as a bigger hit on penetration tests than `unsafe-eval` , so it would be a big win in getting Filament adopted in larger organisations.

**Reply by danharrin** [reply]
Looks perfect to me!

I think we'd need an additional FilamentCspManager and FilamentCsp facade in the support package for those not using the panels. And then `->cspNonce()` on the panel just calls that facade when the panel is booted

**Reply by rossbearman** [reply]
Rather than a new manager and facade, would it make more sense to add these methods to `AssetManager`, as they'll always be used within that context? And update the `HasAssets` trait for the panel to handle the `->cspNonce()` method and property and register it with the `AssetManager` when `registerAssets()` is called?

Someone not using panels would then use `@filamentScripts(['nonce' => $nonce])` to initialise it.

**Reply by danharrin** [reply]
Since it's used alongside inline scripts/styles as well though, is that strictly asset related? I think the line blurs a bit

**Reply by rossbearman** [reply]
After hacking away at this for a while, it doesn't look like it will be possible in Filament as it stands, supporting the inline `<script>` and `<style>` tags is straightforward, but Filament uses `style` attributes and inline event handlers pretty widely and the only solution for those is `unsafe-hashes`, which requires sending a SHA-256 hashed version of each of those attributes with the headers. On top of being a maintenance nightmare, browser support for `unsafe-hashes` is relatively recent.

**Reply by danharrin** [reply]
To be honest a lot of our style attribute usage is abstracted away into helper functions already, is it really that widespread across many views? For example `get_color_css_variables()`

Could create a replacement @style directive that creates the hash attribute at the same time maybe?

**Reply by romansh** [reply]
> On top of being a maintenance nightmare, browser support for `unsafe-hashes` is relatively recent.

Due to 'caniuse' the unsafe-eval usage is less with 92.27%, when for unsafe-hashes it's 94.37%.

**Reply by rossbearman** [reply]
The [`unsafe-eval`](https://caniuse.com/?search=unsafe-eval) stats displayed on caniuse are bugged, they're instead showing the stats for [`wasm-unsafe-eval`](https://caniuse.com/?search=wasm-unsafe-eval), which is much newer. `unsafe-eval` has been in the CSP spec since v1, so any browsers supporting CSP will respect it, and allowing `unsafe-eval` is the only way to use Livewire with a CSP, and that's unlikely to change. 

`unsafe-hashes` was added in CSP v3, but the bigger issue is that you'd have to create a separate SHA-256 hash of every single style attribute and inline event handler used on a page (or that could be passed to the page during a Livewire update), and send those with the initial CSP header.

**Reply by rossbearman** [reply]
@danharrin This is what gets blocked on the default dashboard, once all `<style>` and `<script>` tags have nonces. Each of these that differ would need a SHA-256 hash sent with the initial header, along with a hash for anything that could later get inserted by a Livewire update.

Some of these could potentially be moved into inline `<style>` tags, or moved into the theme style sheet (`scrollbar-gutter: stable` on the sidebar for instance).

![image](https://github.com/filamentphp/filament/assets/212036/9633bc65-14b1-4c33-a5ea-008dd22eb49d)

Resource pages have a few hundred, but I imagine a lot of that is just repetition of table rows or similar.

**Reply by danharrin** [reply]
The only thing we need inline styles for are to pass values (numbers, colors) into CSS props. Any alternatives, maybe ones to use attribute values within props?

**Reply by matthewdanepeavoy** [reply]
Full disclosure, very new to the topic of CSP's, but found something which might be helpful.

@danharrin - Looks like there are two different CSP directives for inline styles.

**style-src**: specifies valid sources for inline styles (<style>, <link rel="stylesheet">). These would be relatively easy to add a nonce too and go a long way towards securing inline <script> and <style> tags.

**style-src-attr**: This is specific only to style="" attributes. 
@rossbearman did not have a directive for this set according to the error he showed, so it was defaulting to "style-src", which requires a nonce. Perhaps style-src-attr: 'self' would work instead? 

A couple of potential options I see from this
**Option 1:**
Use style-src-attr to allow those few attributes, while requiring a nonce for any <style> and <script> tags.

_Ex. 'Content-Security-Policy' => "style-src 'self' 'unsafe-eval' https://fonts.bunny.net 'nonce-adgjdkfgwkejgldfkjgwejg'; style-src-attr 'self'"_

**Option 2:**
Move the attribute values to an inline <style> element (with nonce) as @rossbearman mentioned. This would remove the need for style-src-attr

**Option 3**
Set the css variables through JS. Using .setAttribute() will still trigger the block, but direct style modifying (ex. document.getElementById('1').style.background = 'red') won't.

Alpine already requires "script-src 'unsafe-eval'" so setAttribute might also be a solution.

**Reply by aSeriousDeveloper** [reply]
FYI, Alpine JS now supports `unsafe-eval` a little more! https://github.com/alpinejs/alpine/pull/4671

Additionally, Livewire has a merged PR (but not released yet) that also implements these new changes: https://github.com/livewire/livewire/pull/9461

I imagine making Filament support this `csp-safe` mode might require some changes to Alpine calls throughout the codebase, but it should be an option soon at least.

**Comment by severfire** [comment]
Hi, I do have issue... that may be related... 
![image](https://github.com/user-attachments/assets/76f9c1be-4b8a-4f4d-9f45-3517b06323e7)
I wonder how could i fix it? its on admin login page in JS console... :-/ could anyone help, please?

**Reply by aSeriousDeveloper** [reply]
FYI, Alpine JS now supports `unsafe-eval` a little more! https://github.com/alpinejs/alpine/pull/4671

Additionally, Livewire has a merged PR (but not released yet) that also implements these new changes: https://github.com/livewire/livewire/pull/9461

I imagine making Filament support this `csp-safe` mode might require some changes to Alpine calls throughout the codebase, but it should be an option soon at least.

EDIT: whoops, sorry severfire, replied to the wrong comment :z

**Comment by pixagraphic** [comment]
I agree, we really need CSP using nonces to work with filament. :(

**Comment by kbhnarola** [comment]
Is there any solution? facing the same error.
<img width="1451" height="534" alt="image" src="https://github.com/user-attachments/assets/02704754-14b3-400e-aa13-7b18fb17201c" />

**Comment by CicerBro** [comment]
I always added all the hashes for Filament into my CSP policy, but now on every freaking page load the hash seems to change? How is that even possible?

**Reply by CicerBro** [reply]
If anyone wants the "solution". I use Filament just for my admin panel, and I also use `Spatie/Csp` in Laravel for handling CSP.

So in my custom CSPPolicy.php I just did this:

```php
    protected function shouldOverrideScriptDirective(): bool
    {
        $overridePaths = [
            'admin',
            'admin/*',
            'horizon',
            'horizon/*',
        ];

        foreach ($overridePaths as $pattern) {
            if (request()->is($pattern)) {
                return true;
            }
        }

        return false;
    }
```

And then in the configure method:

```php
if ($this->shouldOverrideScriptDirective()) {
            // Override script directive for specific paths
            $policy->add(blabla);
} else {
            // Normal script directive for all other paths
            $policy->add(blabla);
}

        // Continue with other directives outside if/else
        $policy->add(Directive::FRAME etc. etc.
```

**Comment by 4513** [comment]
Is there actually any update? The Livewire has the 'CSP-safe' feature merged and published a while ago. It's config to enable the CSP so you should not longer need unsafe-eval (as stated in https://livewire.laravel.com/docs/4.x/csp)

The only problem I now see is the Filament itself. 

<img width="892" height="621" alt="image" src="https://github.com/user-attachments/assets/805bf057-1acb-40d2-b8cc-2eaf3e1a3b83" />

(Probably) every reported error leads to Filemant's blades, such as vendor/filament/filament/resources/views/components/theme-switcher/index.blade.php

**Reply by danharrin** [reply]
There's no update on our side, you can help by submitting PRs to make more of our JS CSP-safe though.

**Reply by 4513** [reply]
@danharrin I am willing to help and implement the feature. But I must note, that I am not sure about how much time I can spend in near future to doing so. Anyway, the implementation would consist of several steps because it may, and I believe that it will, impact both, multiple existing features of the filament and filament parties (filament-plugin developers, filament developers, optionally filament users). 

Are you fine about me creating a WIP PR for this, so you and others may all react to the changes and suggest ones? Or is there another preferred way doing so?

Some (if not all) points I will focus on step by step:
* initializing filament with CSP
  * add a Filament PHP API to set nonce for current request - *this is similar to CSP implementation of Vite, Livewire; this should mainly be for filament-user*
  * add a Filament PHP API to render a CSP nonce for script and style HTML elements - *this is very important for everybody creating blade templates and may impact filament-user if creating own templates; should impact filament-plugin-developer if introducing templates with styles/scripts*
* current style/script elements update
  * adding CSP nonces to all existing filament style and script elements
* simple inlines to elements
  * transform current filament's inline styles and scripts into own style/script elements with the available CSP
* livewire/alpine's custom element' data attributes
  *  some attributes currently are written in a way the CSP has problem with that. It usually contains script. The point is to rewrite that attribute's value into separated script element.
* AJAX expecting CSP
  * FIlament may request backend to return a new HTML (e.g. error page on an action). Because the new HTML is not aware of the CSP nonce, it renders with different one, making any styles and scripts of the new HTML being blocked. Not sure right now whether the Livewire has that solved.
  * point of this is to add a CSP-aware property to request body (such as in  POST `/livewire/update` by intercepting) saying what CSP nonce can be parsed. I will have to think a bit more and that part tho, because it might be tricky.
* CSP check TestCase for developers?
  * I will try to create a test case for a developer (mainly filament dev and filament-plugin dev) that would check that all templates are CSP compliant
* filament documentation 
  * update docs for filament-users how to set up filament to comply CSP
  * update docs for developers how to set their plugins CSP compliant and how to test that. That includes how to add CSP nonces to their styles and scripts

I will try to test as much as I can, but honestly, another hand will be really helpful then.

**Reply by danharrin** [reply]
I will note that I don't think Livewire / Alpine / TipTap etc have support for CSP that prevents inline styles. In the Livewire example CSP config, there is `style-src 'self' 'unsafe-inline'`.

So yes, I would accept a WIP PR to start adding CSP support. It will need to be opt-in.

**Comment by m-develops** [comment]
@danharrin I'm interested in working on this, and would want to do it in small
pieces rather than the all-in-one approach in #19230. Before going further,
does this direction look right to you?

**Where I'd start, nonces only:** a closure-based nonce API resolved once per
request, nonces on the asset manager's script output so `@filamentScripts` emits
them (this also covers #8329), and nonces on the 13 inline `<script>` tags in
Filament's own Blade templates. No style changes, since I've read your feedback
on #19230. No Alpine rewrites at this stage. With no nonce configured, output
would be byte-identical to today.

**Two questions:**

1. **Branch.** #19230 targeted `4.x`. New features go to `5.x` now, I assume?

2. **API placement.** `Filament::useCspNonce()` is the obvious name, but
`Filament` is the panels facade and assets render without a panel (the
scaffolding stub layout uses `@filamentScripts` in a plain Laravel layout). So
I'd put it on the support package:

```php
FilamentAsset::useCspNonce(static fn (): string => csp_nonce());
```

plus a `\Filament\Support\csp_nonce()` helper for Blade. A
`Filament::useCspNonce()` passthrough could still be the documented entry point
if you'd prefer. Which would you rather?

**On Alpine, since that seemed to be the blocker on #19230:** Livewire depends
on `@alpinejs/csp`, whose parser does support method calls. I'd assumed from the
docs that it didn't, so the damage is smaller than it looks. Of 312 evaluated
Alpine attributes in `5.x`, 29 fail the grammar: mostly `?.` (13) and `??` (12),
then template literals, arrow functions, `function` and `new`. Concentrated in
`support` / `tables` / `panels` with heavy duplication, so around 10 distinct
problems. `x-html` is also a no-op in the CSP build, used in 2 files. Separate
work, only worth looking at once the nonce part is settled.

**Known limits:** render hooks, `@stack('scripts')` and `RawJs` inject content
Filament can't nonce, and `style-src 'unsafe-inline'` stays required regardless.

@4513, are you still working on this? I offered to split it in March; happy to
leave it with you or collaborate.

**Reply by danharrin** [reply]
1. 5.x please, while 4.x is actively maintained with new features, I think we need Livewire 4 for this
2. I think we should introduce a new facade / manager like `FilamentCsp`? But I don't mind `FilamentAssets` either, whichever you think is more appropriate. If there are gonna be enough non-asset config options required then we should use `FilamentCsp`, but maybe CSP itself is always asset related.

**Comment by m-develops** [comment]
@danharrin phase 1 is up as #20301, ready for review whenever you have time.

One process question, so I do not waste your review time. Would you rather land
that one first and have me open each follow-up only after the previous one
merges, or should I open them as I finish them?

I ask because follow-ups have to target `5.x`, so until #20301 merges their
diffs would contain it as well as their own changes. I am happy either way, I
would just rather not put three overlapping PRs in your queue if you would
prefer to take them one at a time.

For reference, the next one is small: nonces on the 12 inline `<script>` tags in
Filament's own Blade templates, across 6 files. That is what removes the last
violation on a standard panel page. The Alpine work after that does not depend
on the nonce API and could go on its own, though I would want to agree the
approach with you here first, since it changes what plugin authors can write in
their templates.

**Reply by danharrin** [reply]
I can only work through reviews so fast, please wait for one PR to merge before attempting another. Thanks

**Comment by m-develops** [comment]
@danharrin with #20301 approved, the next step is getting a panel to run under Livewire's `csp_safe` mode, so `'unsafe-eval'` can go too. I have prototyped it locally, package by package, and it works: under `csp_safe` a panel's login, dashboard, resource list, search, edit page and action modals run with zero console errors, and behaviour in the default build is identical. Before I turn that into PRs I have three questions, because two of them decide what plugin authors may write in their templates.

The CSP build's evaluator is far stricter than the docs suggest (one expression per attribute, no `if`, no arrow functions, no `?.`/`??`, no globals, `x-html` is a no-op), and Filament had 132 attributes in 44 files that hit one of those. The fix is the one Livewire's own CSP docs recommend, logic out of the attributes and into the packages' existing `Alpine.data()` components, and the PRs will show it better than I can describe it here.

Two Livewire fixes come first, both up as PRs with failing tests: livewire/livewire#10722 (nonce lost on history navigation in SPA mode) and livewire/livewire#10723 (`@script` blocks cannot run under `csp_safe` at all). Filament's six `@script` blocks depend on the second, so no Filament PR before it lands. Of the Alpine plugins Filament loads, floating-ui and focus trapping run clean under the CSP build in the browser; tooltip and sortable I will exercise before opening anything.

**Question 1: arrays in Alpine attributes**

57 of the 132 are `@js($array)` inside an Alpine attribute. `Js::from()` emits `JSON.parse('…')` for arrays, and `JSON` is a global. I checked the parser directly: it accepts a JSON document as an object or array literal, `\u0022` escapes included. So the fix is a helper that emits `json_encode($value, JSON_HEX_*)` inline where an array meets an Alpine attribute, and none of those 57 templates needs restructuring. The alternative, passing data through `data-` attributes and reading `$el.dataset` in the component, works too (the table's group checkboxes do that now) but touches every call site.

I would go with the helper. It also becomes the one rule for plugin authors: in Alpine attributes, keep expressions simple, put logic in `Alpine.data()`, and pass arrays with the helper rather than `@js()`. That is the "maintainability" question from your earlier answer, so I would rather have your view before it spreads.

**Question 2: `x-html`**

`x-html` is a no-op in the CSP build, and Filament uses it in five places (column manager labels, the deferred badge in tabs and dropdown items). A component method setting `innerHTML` from `x-effect` turned out not to survive a Livewire morph, so the workable option is a server-side branch on `Livewire::isCspSafe()` that renders `x-text` with a plain-text label in the CSP build. Cost: `Htmlable` labels lose their markup there. If you would rather keep templates free of that branch, the alternative is plain-text labels in those five places in both builds.

**Question 3: the three features that run JavaScript from a string**

`afterStateUpdatedJs()`, `JsContent` (`Text::make()->js()`) and `RawJs` run user-supplied JavaScript from a PHP string. That is `eval()` by construction, and no rewrite gets around it: without `'unsafe-eval'` the string cannot become code. Under `csp_safe` they fail silently today, one console error each, and Filament uses them itself in the rich editor's grid action, in `Text`, and in the slider's `format()` and `filters()`.

The need behind them is fine under CSP as long as the code lives in a bundle instead of a string. So I would like to propose an opt-in alternative rather than only documenting the hole: the developer registers a function once from their own (nonced) script, and refers to it by name.

```php
->afterStateUpdatedJs(handler: 'orders.recalculateTotal')
```

```js
Filament.registerJs('orders.recalculateTotal', ($state, $old) => { … })
```

Filament's side becomes `$wire.watch(path, filamentJs('orders.recalculateTotal'))`, which the CSP parser accepts. The same shape works for `JsContent` and `RawJs`. The string forms stay exactly as they are for everyone else; the docs say that under `csp_safe` you register functions instead. The three in-tree usages would move to the registry too, so Filament's own components keep working. I think this is the nicer API even outside CSP, but it is additive surface, so I would rather have your answer before building it. Until then those three stay documented as unsupported under `csp_safe`.

**How I would split it**

One PR per package, each complete, each a no-op in the default build: `tables`, `support` (the helper and the shared bits), `schemas`, `actions`, `panels`, `forms`, `widgets`, `notifications`, plus a docs page. Only someone who enables `csp_safe` sees any difference, and only fully once the last one is in. `style-src 'unsafe-inline'` stays required as agreed before.

**Reply by danharrin** [reply]
Thanks for the detailed investigation. I’ve looked further into how Alpine and Livewire implement CSP-safe mode, and I think we should move more of this upstream before starting the Filament package PRs.

**1. Arrays in Alpine attributes**

I would rather not introduce and maintain a Filament-specific replacement for Laravel’s `@js()`. `@js()` is common and already safe inside attributes; the incompatibility is that Alpine’s CSP evaluator does not expose the narrowly required `JSON.parse()`.

Could you pursue focused Alpine support for Laravel’s existing `JSON.parse('...')` output? Ideally this would be a safe internal intrinsic rather than exposing the native `JSON` global generally.

While working upstream, I think it is also reasonable to add the common low-risk syntax that Filament needs: `??`, `?.`, and semicolon-separated expression sequences. I would not try to add arrows, statements, `new`, or unrestricted globals.

**2. `x-html`**

Setting `innerHTML` is not itself prohibited by CSP; Alpine disables `x-html` as an additional sandboxing decision.

For Filament’s server-generated labels and icons, I would prefer us to add a Filament-specific directive such as `x-filament-html` rather than dropping the markup. It should only be used for trusted or sanitized framework HTML and should not initialize additional Alpine directives inside the inserted content unless we discover that is necessary.

**3. JavaScript string APIs**

I do not think we need the named function registry initially. Alpine’s CSP evaluator can parse and interpret a string without using browser `eval()`.

We should keep APIs such as `afterStateUpdatedJs()` and evaluate their expressions through Alpine’s CSP evaluator from Filament’s bundled component code. For example, the bundled component can create the `$wire.watch()` callback and evaluate the supplied expression with `$state`, `$old`, `$get()` and `$set()` in scope.

The same approach should work for `JsContent` and some `RawJs` use cases. Arbitrary JavaScript programs, inline callback definitions and complex objects containing functions can remain documented as unsupported under CSP-safe mode.

Please pursue the focused Alpine changes first. Once those dependencies are settled, we can continue with one Filament PR at a time as previously discussed.

**Reply by m-develops** [reply]
@danharrin the three Alpine PRs are up, each on its own, with the motivation and a link to your answer in the description:

- alpinejs/alpine#4913: `JSON.parse()` with a string as an intrinsic, so `@js()` output works unchanged. The rest of `JSON` stays unreachable.
- alpinejs/alpine#4914: `?.` and `??`, with whole-chain short-circuiting and lazy evaluation of the right side.
- alpinejs/alpine#4915: semicolon-separated expressions, evaluated in order, last value returned.

Once those are through I will pick up the Filament side the way you described: `x-filament-html` in support, `afterStateUpdatedJs()` / `JsContent` / `RawJs` evaluated through the CSP evaluator from bundled code with `$state`, `$old`, `$get()` and `$set()` in scope, then one package PR at a time after #20301 and livewire/livewire#10723 land.
